openedx/core/djangoapps/content_libraries/libraries_index.py

```diff
@@ -0,0 +1,159 @@
+""" Code to allow indexing content libraries """
+
+import logging
+
+from django.conf import settings
+from django.dispatch import receiver
+from elasticsearch.exceptions import ConnectionError
+from search.search_engine_base import SearchEngine
+
+from openedx.core.djangoapps.content_libraries.signals import (
+    CONTENT_LIBRARY_CREATED,
+    CONTENT_LIBRARY_UPDATED,
+    CONTENT_LIBRARY_DELETED,
+    LIBRARY_BLOCK_CREATED,
+    LIBRARY_BLOCK_UPDATED,
+    LIBRARY_BLOCK_DELETED,
+)
+from openedx.core.djangoapps.content_libraries.library_bundle import LibraryBundle
+from openedx.core.djangoapps.content_libraries.models import ContentLibrary
+from openedx.core.lib.blockstore_api import get_bundle
+
+log = logging.getLogger(__name__)
+
+MAX_SIZE = 10000  # 10000 is the maximum records elastic is able to return in a single result. Defaults to 10.
+
+
+class LibraryNotIndexedException(Exception):
+    """
+    Library supplied wasn't indexed in ElasticSearch
+    """
+    pass
+
+
+class ContentLibraryIndexer:
+    """
+    Class to perform indexing for blockstore-based content libraries
+    """
+
+    INDEX_NAME = "content_library_index"
+    LIBRARY_DOCUMENT_TYPE = "content_library"
+
+    @classmethod
+    def index_libraries(cls, library_keys):
+        """
+        Index the specified libraries. If they already exist, replace them with new ones.
+        """
+        from openedx.core.djangoapps.content_libraries.api import DRAFT_NAME
+
+        searcher = SearchEngine.get_search_engine(cls.INDEX_NAME)
+
+        library_dicts = []
+
+        for library_key in library_keys:
+            ref = ContentLibrary.objects.get_by_key(library_key)
+            lib_bundle = LibraryBundle(library_key, ref.bundle_uuid, draft_name=DRAFT_NAME)
+            num_blocks = len(lib_bundle.get_top_level_usages())
+            last_published = lib_bundle.get_last_published_time()
+            last_published_str = None
+            if last_published:
+                last_published_str = last_published.strftime('%Y-%m-%dT%H:%M:%SZ')
+            (has_unpublished_changes, has_unpublished_deletes) = lib_bundle.has_changes()
+
+            bundle_metadata = get_bundle(ref.bundle_uuid)
+
+            library_dict = {
+                "id": str(library_key),
+                "uuid": str(bundle_metadata.uuid),
+                "title": bundle_metadata.title,
+                "description": bundle_metadata.description,
+                "num_blocks": num_blocks,
+                "version": bundle_metadata.latest_version,
+                "last_published": last_published_str,
+                "has_unpublished_changes": has_unpublished_changes,
+                "has_unpublished_deletes": has_unpublished_deletes,
+            }
```

Comment:
This is slightly complicated - it will require a reindex of libraries in the current design. Which means that changes which update indexes will come with instructions to reindex libraries.

There's probably a way to do this through migrations, but am not sure if that's a good idea. I'm open to suggestions here though.

Related conversation from our ticket (it was concerning only the initial index, so migrations might not be a very bad idea for index updates):

> > I had a compatibility question for indexing as well - when the code will be first deployed, existing libraries won't have any existing indexes. We can start populating them as needed (might be slightly complicated when accounting for possible failures), or have a way to execute mass-indexing when upgrade happens. Do you have any thoughts on this, or is there any existing common method which is used in doing such one-time executions?

> It's usual practice to provide a management command that will re-index either a specific library or all libraries. Then it's up to the ops team for each edX instance to decide when to run it. In this case in practice the only significant installation of Open edX that has any content libraries in the new format/blockstore is courses.edx.org (though it does have literally thousands of v2 libraries, because every content item on LabXchange is its own blockstore content library).
See https://github.com/edx/edx-platform/blob/master/cms/djangoapps/contentstore/management/commands/reindex_library.py , https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/content/learning_sequences/management/commands/update_course_outline.py , etc.
It can also be done as a data migration in a django migration file, but this gives the ops team less control and less visibility so is not recommended.

Reply:
(@ormsbee , feel free to chime in here)

Re. the conversation on the ticket: yes, having a management command to bulk-reindex is an established pattern in the platform. When a command like that gets too heavy for edx.org, we often turn it into a "batch" command that reindexes some subset of the items, and then write a cron job to regularly runs that batch command as long as there are items that need reindexing. Let's stick with the all-at-once command you have, though, at least for now.

> This is slightly complicated - it will require a reindex of libraries in the current design. Which means that changes which update indexes will come with instructions to reindex libraries.

This makes sense. However, it we be good if schema changes were more "fail-safe", such that if an operator waits too long or simply doesn't think to run the re-index job, we fall back to the minimal listing behavior instead of risking the `AttributeError`s, `KeyError`s, etc. that I imagine could result from using outdated cache entries.

IMO most obvious way of doing this (which is also established in edx-platform) is by adding version number to the schema. It would start at 0 in this PR, with a big comment telling devs to bump the version number if they change the schema. Then, refuse to load any cache entries with a different version number.

Alternatively, you could add some validation (do we keys X, Y, Z? are they non-null? etc.) after loading the cached library, and refuse to load the library if it fails those checks, falling back to the minimal listing. Then, any breaking schema changes should cause the validation to fail.